flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/ZooKeeperCompletedCheckpointStoreTest.java

```diff
@@ -160,9 +162,12 @@ public Void answer(InvocationOnMock invocation) throws Throwable {
 			stateStorage,
 			Executors.directExecutor());
 
-		SharedStateRegistry sharedStateRegistry = new SharedStateRegistry();
+		SharedStateRegistry sharedStateRegistry = spy(new SharedStateRegistry());
 		zooKeeperCompletedCheckpointStore.recover(sharedStateRegistry);
 
+		verify(retrievableStateHandle1.retrieveState(), times(1)).registerSharedStatesAfterRestored(sharedStateRegistry);
```

Comment:
The behaviour that is tested here is not strongly connected to the bug, I just noticed it might be also worth checking to avoid future problems. {{StandaloneCompletedCheckpointStore}} is not expected to do this, so there is no need for a this particular test.

The question for a general test is valid. However, I noticed that externalized checkpoints in general are not tested at all, so this is not just a small addition but a whole new test (including some required test methods to trigger and restore from externalized checkpoints in the JM) that would almost justify a new planning task and PR.